Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift

```diff
@@ -0,0 +1,634 @@
+            self.stats.availableStreams -= closedAction.maxStreams - closedAction.usedStreams
+
+            switch closedAction.previousConnectionState {
+            case .idle:
+                self.stats.idle -= 1
+
+            case .leased:
+                self.stats.leased -= 1
+
+            case .closing:
+                self.stats.closing -= 1
+            }
+
+            let lastIndex = self.connections.endIndex - 1
+
+            if index == lastIndex {
+                self.connections.remove(at: index)
+            } else {
+                self.connections.swapAt(index, lastIndex)
+                self.connections.remove(at: lastIndex)
+            }
+
+            return FailedConnectionContext(connectionsStarting: 0)
+        }
+
+        // MARK: Shutdown
+
+        mutating func triggerShutdown(_ cleanup: inout ConnectionAction.Shutdown) {
+            fatalError("TODO: Unimplemented")
+//            for var connectionState in self.connections {
+//                let connectionCleanup = connectionState.shutdown()
+//                if let connection = connectionCleanup.connection {
+//                    cleanup.connections.append(connection)
+//                }
+//                cleanup.timersToCancel.append(contentsOf: connectionCleanup.timersToCancel)
+//            }
+//
+//            self.connections = []
+        }
+
+        // MARK: - Private functions -
+
+        @usableFromInline
+        /*private*/ func getConnectionUse(index: Int) -> ConnectionUse {
+            switch index {
+            case 0..<self.minimumConcurrentConnections:
+                return .persisted
+            case self.minimumConcurrentConnections..<self.maximumConcurrentConnectionSoftLimit:
+                return .demand
+            case self.maximumConcurrentConnectionSoftLimit...:
+                return .overflow
+            default:
+                preconditionFailure()
+            }
+        }
+
+        @usableFromInline
+        /*private*/ func makeAvailableConnectionContextForConnection(at index: Int, info: ConnectionAvailableInfo) -> AvailableConnectionContext {
+            precondition(self.connections[index].isAvailable)
+            let use = self.getConnectionUse(index: index)
+            return AvailableConnectionContext(use: use, info: info)
+        }
+
+        @inlinable
+        /*private*/ func findAvailableConnection() -> Int? {
+            return self.connections.firstIndex(where: { $0.isAvailable })
+        }
+
+        @inlinable
+        /*private*/ mutating func swapForDeletion(index indexToDelete: Int) -> TimerCancellationToken? {
+            let lastConnectedIndex = self.connections.lastIndex(where: { $0.isConnected })
+
+            if lastConnectedIndex == nil || lastConnectedIndex! < indexToDelete {
+                self.removeO1(indexToDelete)
+                return nil
+            }
+
+            guard let lastConnectedIndex = lastConnectedIndex else { preconditionFailure() }
```

Comment:
```suggestion
            guard let lastConnectedIndex else { preconditionFailure() }
```
Also, with no message you might as well just use `!` for all the difference it makes.